File: sdk/pythonexamples/transaction_multisig.py
#!/usr/bin/env python

#
# Shows how to create multisig account.
#

from binascii import unhexlify

import sha3

from symbolchain.core.CryptoTypes import Hash256, PrivateKey
from symbolchain.core.facade.SymFacade import SymFacade
from symbolchain.core.sym.KeyPair import KeyPair
from symbolchain.core.sym.MerkleHashBuilder import MerkleHashBuilder


class MultisigAccountModificationSample:
    def __init__(self):
        self.facade = SymFacade('public_test')
        self.multisig_key_pair = KeyPair(PrivateKey(unhexlify('11002233445566778899AABBCCDDEEFF11002233445566778899AABBCCDDEEFF')))
        self.cosignatory_key_pairs = [
            KeyPair(PrivateKey(unhexlify('AABBCCDDEEFF11002233445566778899AABBCCDDEEFF11002233445566778899'))),
            KeyPair(PrivateKey(unhexlify('BBCCDDEEFF11002233445566778899AABBCCDDEEFF11002233445566778899AA'))),
            KeyPair(PrivateKey(unhexlify('CCDDEEFF11002233445566778899AABBCCDDEEFF11002233445566778899AABB')))
        ]

    def run(self):
        # note: it's important to SIGN the transaction BEFORE adding cosignatures
        aggregate_transaction = self.create_aggregate_transaction()

        signature = self.facade.sign_transaction(self.multisig_key_pair, aggregate_transaction)
        self.facade.transaction_factory.attach_signature(aggregate_transaction, signature)

        self.add_cosignatures(aggregate_transaction)

        print(aggregate_transaction)

    def create_aggregate_transaction(self):
        embedded_transactions = [
            self.facade.transaction_factory.create_embedded({
                'type': 'multisigAccountModification',
                'signer_public_key': self.multisig_key_pair.public_key,
                'min_approval_delta': 1,
                'min_removal_delta': 1,
                'address_additions': list(map(self.to_address, self.cosignatory_key_pairs))
            })
        ]

        return self.facade.transaction_factory.create({
            'type': 'aggregateComplete',
            'signer_public_key': self.multisig_key_pair.public_key,
            'fee': 625,
            'deadline': 12345,
            'transactions_hash': self.calculate_transactions_hash(embedded_transactions).bytes,
            'transactions': embedded_transactions,
        })

    @staticmethod
    def calculate_transactions_hash(transactions):
        hash_builder = MerkleHashBuilder()
        for embedded_transaction in transactions:
            hash_builder.update(Hash256(sha3.sha3_256(embedded_transaction.serialize()).digest()))
        return hash_builder.final()

    def to_address(self, key_pair):
        return self.facade.network.public_key_to_address(key_pair.public_key)

    def add_cosignatures(self, aggregate_transaction):
        transaction_hash = self.facade.hash_transaction(aggregate_transaction).bytes
        for key_pair in self.cosignatory_key_pairs:
            cosignature = (0, key_pair.public_key.bytes, key_pair.sign(transaction_hash).bytes)
            aggregate_transaction.cosignatures.append(cosignature)


def main():
    sample = MultisigAccountModificationSample()
    sample.run()


if __name__ == '__main__':
    main()

File: sdk/java/examples/src/main/java/org/symbol/examples/TransactionMultisig.java
package org.symbol.examples;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

import org.symbol.sdk.CryptoTypes;
import org.symbol.sdk.facade.SymbolFacade;
import org.symbol.sdk.symbol.Address;
import org.symbol.sdk.symbol.KeyPair;
import org.symbol.sdk.symbol.SymbolTransactionFactory;
import org.symbol.sdk.symbol.descriptors.AggregateCompleteTransactionV3Descriptor;
import org.symbol.sdk.symbol.descriptors.MultisigAccountModificationTransactionV1Descriptor;
import org.symbol.sdk.symbol.models.AggregateCompleteTransactionV3;
import org.symbol.sdk.symbol.models.Cosignature;
import org.symbol.sdk.symbol.models.EmbeddedTransaction;

/**
 * Creates a multisig account: builds an aggregate-complete transaction wrapping a single
 * {@code multisig_account_modification_transaction_v1}, signs it with the multisig key, then attaches three cosignatures.
 */
public final class TransactionMultisig {
	private TransactionMultisig() {
	}

	private static KeyPair createKeyPair(final String privateKeyHex) {
		return new KeyPair(new CryptoTypes.PrivateKey(privateKeyHex));
	}

	private final SymbolFacade facade = new SymbolFacade("testnet");

	private final KeyPair multisigKeyPair = createKeyPair("11002233445566778899AABBCCDDEEFF11002233445566778899AABBCCDDEEFF");

	private final List<KeyPair> cosignatoryKeyPairs = List.of(
			createKeyPair("AABBCCDDEEFF11002233445566778899AABBCCDDEEFF11002233445566778899"),
			createKeyPair("BBCCDDEEFF11002233445566778899AABBCCDDEEFF11002233445566778899AA"),
			createKeyPair("CCDDEEFF11002233445566778899AABBCCDDEEFF11002233445566778899AABB"));

	private AggregateCompleteTransactionV3 createAggregateTransaction() {
		final List<Address> addressAdditions = new ArrayList<>();
		for (final KeyPair cosigner : cosignatoryKeyPairs)
			addressAdditions.add(facade.network.publicKeyToAddress(cosigner.getPublicKey()));

		final MultisigAccountModificationTransactionV1Descriptor embeddedDescriptor = new MultisigAccountModificationTransactionV1Descriptor(
				1, 1, addressAdditions, (List<Address>) null);
		final List<EmbeddedTransaction> embeddedTransactions = List
				.of(facade.createEmbeddedTransactionFromTypedDescriptor(embeddedDescriptor, multisigKeyPair.getPublicKey()));

		final AggregateCompleteTransactionV3Descriptor aggregateDescriptor = new AggregateCompleteTransactionV3Descriptor(
				SymbolFacade.hashEmbeddedTransactions(embeddedTransactions), embeddedTransactions, (List<Cosignature>) null);

		// pin deterministic header values in the descriptor (like the JS / Python examples) so the output stays comparable
		final Map<String, Object> aggregateMap = aggregateDescriptor.toMap();
		aggregateMap.put("signerPublicKey", multisigKeyPair.getPublicKey());
		aggregateMap.put("fee", 625L);
		aggregateMap.put("deadline", 12345L);
		return (AggregateCompleteTransactionV3) facade.transactionFactory.create(aggregateMap);
	}

	private void addCosignatures(final AggregateCompleteTransactionV3 aggregateTransaction) {
		final List<Cosignature> cosignatures = new ArrayList<>();
		for (final KeyPair keyPair : cosignatoryKeyPairs)
			cosignatures.add(facade.cosignTransaction(keyPair, aggregateTransaction));

		aggregateTransaction.setCosignatures(cosignatures);
	}

	private void run() {
		// note: it's important to SIGN the transaction BEFORE adding cosignatures
		final AggregateCompleteTransactionV3 aggregateTransaction = createAggregateTransaction();

		final CryptoTypes.Signature signature = facade.signTransaction(multisigKeyPair, aggregateTransaction);
		SymbolTransactionFactory.attachSignature(aggregateTransaction, signature);

		System.out.println("Hash: " + facade.hashTransaction(aggregateTransaction));

		addCosignatures(aggregateTransaction);

		System.out.println("Cosignatures: " + aggregateTransaction.getCosignatures().size() + "\n");
		System.out.println(aggregateTransaction.toString());
	}

	public static void main(final String[] args) {
		new TransactionMultisig().run();
	}
}
